File: src/spencer/provider.py
from __future__ import annotations

import json
import logging
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Callable

from .config import Settings


LOGGER = logging.getLogger("spencer.provider")


class ProviderError(RuntimeError):
    """Raised when Spencer cannot complete a provider request."""


class RetryableProviderError(ProviderError):
    """Raised for provider failures that are safe to retry."""


@dataclass(frozen=True)
class ProviderToolFunction:
    name: str
    arguments: str


@dataclass(frozen=True)
class ProviderToolCall:
    id: str
    function: ProviderToolFunction


@dataclass(frozen=True)
class ProviderMessage:
    content: str | None
    tool_calls: list[ProviderToolCall]


@dataclass(frozen=True)
class ProviderChoice:
    message: ProviderMessage


@dataclass(frozen=True)
class ProviderResponse:
    choices: list[ProviderChoice]


Transport = Callable[[urllib.request.Request, int], dict[str, Any]]


class HTTPProvider:
    """Provider-neutral HTTP client with configurable protocol normalization."""

    def __init__(self, settings: Settings, *, transport: Transport | None = None, retries: int = 2):
        self.settings = settings
        self.retries = max(0, retries)
        self.transport = transport or self._request_json

    def complete(
        self,
        *,
        model: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
    ) -> ProviderResponse:
        if not self.settings.api_url:
            raise ProviderError(
                "No API URL configured. Set SPENCER_API_URL or configure api_url in .spencer.toml."
            )
        payload = self._build_payload(model=model, messages=messages, tools=tools)
        headers = {"Content-Type": "application/json", **self.settings.headers}
        if self.settings.api_key and self.settings.api_key_header not in headers:
            value = self.settings.api_key
            if self.settings.api_key_prefix:
                value = f"{self.settings.api_key_prefix} {value}"
            headers[self.settings.api_key_header] = value
        request = urllib.request.Request(
            self.settings.api_url,
            data=json.dumps(payload).encode("utf-8"),
            headers=headers,
            method="POST",
        )

        last_error: Exception | None = None
        for attempt in range(self.retries + 1):
            try:
                started = time.monotonic()
                raw = self.transport(request, self.settings.api_timeout)
                response = self._normalize_response(raw)
                LOGGER.debug("provider request succeeded in %.2fs", time.monotonic() - started)
                return response
            except Exception as exc:
                last_error = exc
                retryable = isinstance(exc, (RetryableProviderError, urllib.error.URLError, TimeoutError))
                if not retryable or attempt >= self.retries:
                    break
                delay = 2**attempt
                LOGGER.warning("provider request failed (%s); retrying in %ss", type(exc).__name__, delay)
                time.sleep(delay)
        raise ProviderError(f"Provider request failed after {self.retries + 1} attempts: {last_error}") from last_error

    def _build_payload(
        self,
        *,
        model: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
    ) -> dict[str, Any]:
        protocol = self.settings.protocol
        if protocol == "anthropic-messages":
            system = "\n\n".join(
                str(message["content"])
                for message in messages
                if message.get("role") == "system"
            )
            anthropic_messages = []
            for message in messages:
                role = message.get("role")
                if role == "system":
                    continue
                if role == "tool":
                    anthropic_messages.append(
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "tool_result",
                                    "tool_use_id": message.get("tool_call_id", "unknown"),
                                    "content": str(message.get("content", "")),
                                }
                            ],
                        }
                    )
                elif role == "assistant" and message.get("tool_calls"):
                    blocks = []
                    if message.get("content"):
                        blocks.append({"type": "text", "text": message["content"]})
                    for call in message["tool_calls"]:
                        function = call.get("function", {})
                        try:
                            tool_input = json.loads(function.get("arguments", "{}"))
                        except json.JSONDecodeError:
                            tool_input = {}
                        blocks.append(
                            {
                                "type": "tool_use",
                                "id": call.get("id", "unknown"),
                                "name": function.get("name", ""),
                                "input": tool_input,
                            }
                        )
                    anthropic_messages.append({"role": "assistant", "content": blocks})
                else:
                    anthropic_messages.append({"role": "user" if role == "user" else "assistant", "content": message.get("content", "")})
            payload = {
                **self.settings.request_fields,
                "model": model,
                "system": system,
                "messages": anthropic_messages,
                "max_tokens": 4_500,
                "tools": [
                    {
                        "name": tool["function"]["name"],
                        "description": tool["function"].get("description", ""),
                        "input_schema": tool["function"].get("parameters", {}),
                    }
                    for tool in tools
                ],
            }
            return payload
        if protocol == "ollama-chat":
            return {**self.settings.request_fields, "model": model, "messages": messages, "tools": tools, "stream": False}
        return {
            **self.settings.request_fields,
            "model": model,
            "messages": messages,
            "tools": tools,
            "tool_choice": "auto",
        }

    def _normalize_response(self, data: dict[str, Any]) -> ProviderResponse:
        if self.settings.protocol == "anthropic-messages":
            return self._normalize_anthropic(data)
        if self.settings.protocol == "ollama-chat":
            message = data.get("message", {})
            return ProviderResponse([ProviderChoice(self._message_from_mapping(message))])

        message_data = _value_at_path(data, self.settings.content_path.rsplit(".", 1)[0], {})
        if not isinstance(message_data, dict):
            message_data = {}
        content = _value_at_path(data, self.settings.content_path)
        if content is not None and not isinstance(content, str):
            content = str(content)
        raw_tool_calls = _value_at_path(data, self.settings.tool_calls_path, [])
        if not isinstance(raw_tool_calls, list):
            raw_tool_calls = []
        tool_calls = [self._tool_call_from_mapping(item, index) for index, item in enumerate(raw_tool_calls)]
        if not content and not tool_calls and "error" in data:
            raise ProviderError(f"Provider returned an error: {data['error']}")
        return ProviderResponse([ProviderChoice(ProviderMessage(content=content, tool_calls=tool_calls))])

    def _normalize_anthropic(self, data: dict[str, Any]) -> ProviderResponse:
        content_parts = data.get("content", [])
        text_parts = [part.get("text", "") for part in content_parts if part.get("type") == "text"]
        tool_calls = []
        for part in content_parts:
            if part.get("type") == "tool_use":
                tool_calls.append(
                    ProviderToolCall(
                        id=str(part.get("id", f"tool-{len(tool_calls)}")),
                        function=ProviderToolFunction(
                            name=str(part.get("name", "")),
                            arguments=json.dumps(part.get("input", {})),
                        ),
                    )
                )
        return ProviderResponse([ProviderChoice(ProviderMessage("\n".join(text_parts) or None, tool_calls))])

    def _message_from_mapping(self, message: Any) -> ProviderMessage:
        if not isinstance(message, dict):
            return ProviderMessage(None, [])
        content = message.get("content")
        if content is not None and not isinstance(content, str):
            content = str(content)
        raw_tool_calls = message.get("tool_calls", [])
        if not isinstance(raw_tool_calls, list):
            raw_tool_calls = []
        return ProviderMessage(
            content=content,
            tool_calls=[self._tool_call_from_mapping(item, index) for index, item in enumerate(raw_tool_calls)],
        )

    @staticmethod
    def _tool_call_from_mapping(item: Any, index: int) -> ProviderToolCall:
        if not isinstance(item, dict):
            return ProviderToolCall(f"tool-{index}", ProviderToolFunction("", "{}"))
        function = item.get("function", item)
        if not isinstance(function, dict):
            function = {}
        arguments = function.get("arguments", function.get("input", {}))
        if not isinstance(arguments, str):
            arguments = json.dumps(arguments)
        return ProviderToolCall(
            id=str(item.get("id", f"tool-{index}")),
            function=ProviderToolFunction(str(function.get("name", item.get("name", ""))), arguments),
        )

    @staticmethod
    def _request_json(request: urllib.request.Request, timeout: int) -> dict[str, Any]:
        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                body = response.read()
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")[:1_000]
            error_type = RetryableProviderError if exc.code == 429 or exc.code >= 500 else ProviderError
            raise error_type(f"HTTP {exc.code}: {detail}") from exc
        except urllib.error.URLError as exc:
            raise RetryableProviderError(f"Network error: {exc.reason}") from exc
        try:
            value = json.loads(body.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ProviderError("Provider returned a non-JSON response") from exc
        if not isinstance(value, dict):
            raise ProviderError("Provider response must be a JSON object")
        return value


# Backwards-compatible name for integrations that imported the prototype class.
ChatProvider = HTTPProvider


def _value_at_path(data: Any, path: str, default: Any = None) -> Any:
    current = data
    for component in path.split("."):
        if isinstance(current, list) and component.isdigit():
            index = int(component)
            if index >= len(current):
                return default
            current = current[index]
        elif isinstance(current, dict) and component in current:
            current = current[component]
        else:
            return default
    return current

File: lib/provider.js
'use strict';

const { setTimeout: sleep } = require('node:timers/promises');

const BUILTIN_PROTOCOLS = ['generic-json', 'openai-compatible', 'anthropic-messages', 'ollama-chat'];
const backends = new Map();

class ProviderError extends Error {
  constructor(message, options = {}) {
    super(message, options);
    this.name = 'ProviderError';
  }
}

class RetryableProviderError extends ProviderError {
  constructor(message, options = {}) {
    super(message, options);
    this.name = 'RetryableProviderError';
  }
}

function registerBackend(name, factory) {
  if (!name || typeof name !== 'string') throw new TypeError('Backend name must be a non-empty string');
  if (typeof factory !== 'function') throw new TypeError('Backend factory must be a function');
  backends.set(name, factory);
}

function availableBackends() {
  return [...new Set([...BUILTIN_PROTOCOLS, ...backends.keys()])].sort();
}

function valueAtPath(data, dottedPath, fallback = undefined) {
  let current = data;
  for (const component of String(dottedPath).split('.')) {
    if (Array.isArray(current) && /^\d+$/.test(component)) {
      current = current[Number(component)];
    } else if (current && typeof current === 'object' && component in current) {
      current = current[component];
    } else {
      return fallback;
    }
  }
  return current;
}

function toolCallFromMapping(item, index) {
  if (!item || typeof item !== 'object') return { id: `tool-${index}`, function: { name: '', arguments: '{}' } };
  const fn = item.function && typeof item.function === 'object' ? item.function : item;
  let argumentsValue = fn.arguments ?? fn.input ?? {};
  if (typeof argumentsValue !== 'string') argumentsValue = JSON.stringify(argumentsValue);
  return {
    id: String(item.id ?? `tool-${index}`),
    function: { name: String(fn.name ?? item.name ?? ''), arguments: argumentsValue },
  };
}

function normalizeGeneric(data, settings) {
  const contentValue = valueAtPath(data, settings.contentPath);
  const content = contentValue == null ? null : String(contentValue);
  const rawCalls = valueAtPath(data, settings.toolCallsPath, []);
  const toolCalls = Array.isArray(rawCalls) ? rawCalls.map(toolCallFromMapping) : [];
  if (!content && !toolCalls.length && data && data.error) throw new ProviderError(`Provider returned an error: ${JSON.stringify(data.error)}`);
  return { choices: [{ message: { content, tool_calls: toolCalls } }] };
}

function toAnthropicMessages(messages) {
  const system = messages.filter((message) => message.role === 'system').map((message) => String(message.content ?? '')).join('\n\n');
  const converted = [];
  for (const message of messages) {
    if (message.role === 'system') continue;
    if (message.role === 'tool') {
      converted.push({ role: 'user', content: [{ type: 'tool_result', tool_use_id: message.tool_call_id ?? 'unknown', content: String(message.content ?? '') }] });
    } else if (message.role === 'assistant' && message.tool_calls?.length) {
      const blocks = [];
      if (message.content) blocks.push({ type: 'text', text: message.content });
      for (const call of message.tool_calls) {
        let input = {};
        try { input = JSON.parse(call.function?.arguments ?? '{}'); } catch { /* leave empty */ }
        blocks.push({ type: 'tool_use', id: call.id ?? 'unknown', name: call.function?.name ?? '', input });
      }
      converted.push({ role: 'assistant', content: blocks });
    } else {
      converted.push({ role: message.role === 'user' ? 'user' : 'assistant', content: message.content ?? '' });
    }
  }
  return { system, messages: converted };
}

function anthropicPayload(model, messages, tools, settings) {
  const converted = toAnthropicMessages(messages);
  return {
    ...settings.requestFields,
    model,
    system: converted.system,
    messages: converted.messages,
    max_tokens: 4500,
    tools: tools.map((tool) => ({
      name: tool.function.name,
      description: tool.function.description ?? '',
      input_schema: tool.function.parameters ?? {},
    })),
  };
}

function normalizeAnthropic(data) {
  const blocks = Array.isArray(data?.content) ? data.content : [];
  const text = blocks.filter((block) => block.type === 'text').map((block) => block.text ?? '').join('\n');
  const toolCalls = blocks.filter((block) => block.type === 'tool_use').map((block, index) => ({
    id: String(block.id ?? `tool-${index}`),
    function: { name: String(block.name ?? ''), arguments: JSON.stringify(block.input ?? {}) },
  }));
  return { choices: [{ message: { content: text || null, tool_calls: toolCalls } }] };
}

function createBuiltinBackend(protocol, settings) {
  if (protocol === 'anthropic-messages') {
    return {
      buildPayload: (model, messages, tools) => anthropicPayload(model, messages, tools, settings),
      normalizeResponse: normalizeAnthropic,
    };
  }
  if (protocol === 'ollama-chat') {
    return {
      buildPayload: (model, messages, tools) => ({ ...settings.requestFields, model, messages, tools, stream: false }),
      normalizeResponse: (data) => ({ choices: [{ message: { content: data?.message?.content ?? null, tool_calls: (data?.message?.tool_calls ?? []).map(toolCallFromMapping) } }] }),
    };
  }
  return {
    buildPayload: (model, messages, tools) => ({ ...settings.requestFields, model, messages, tools, tool_choice: 'auto' }),
    normalizeResponse: (data) => normalizeGeneric(data, settings),
  };
}

function backendFor(settings) {
  const customFactory = backends.get(settings.protocol);
  if (customFactory) return customFactory(settings);
  if (BUILTIN_PROTOCOLS.includes(settings.protocol)) return createBuiltinBackend(settings.protocol, settings);
  throw new ProviderError(`Unknown provider protocol '${settings.protocol}'. Available backends: ${availableBackends().join(', ')}`);
}

function authHeaders(settings) {
  const headers = { 'content-type': 'application/json', ...(settings.headers ?? {}) };
  if (settings.apiKey && !headers[settings.apiKeyHeader]) {
    headers[settings.apiKeyHeader] = `${settings.apiKeyPrefix ?? 'Bearer'}${settings.apiKeyPrefix ? ' ' : ''}${settings.apiKey}`;
  }
  return headers;
}

async function requestJson(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(url, { ...options, signal: controller.signal });
    const text = await response.text();
    let body;
    try { body = JSON.parse(text); } catch { throw new ProviderError(`Provider returned non-JSON HTTP ${response.status}`); }
    if (!response.ok) {
      const ErrorType = response.status === 429 || response.status >= 500 ? RetryableProviderError : ProviderError;
      throw new ErrorType(`Provider HTTP ${response.status}: ${JSON.stringify(body).slice(0, 1000)}`);
    }
    return body;
  } catch (error) {
    if (error.name === 'AbortError') throw new RetryableProviderError(`Provider request timed out after ${timeoutMs}ms`);
    if (error instanceof ProviderError) throw error;
    throw new RetryableProviderError(`Provider network error: ${error.message}`);
  } finally {
    clearTimeout(timer);
  }
}

async function complete(settings, { model, messages, tools }, options = {}) {
  if (!settings.apiUrl) throw new ProviderError('No API URL configured. Set SPENCER_API_URL or configure apiUrl.');
  const backend = options.backend ?? backendFor(settings);
  const transport = options.transport ?? requestJson;
  const payload = backend.buildPayload(model, messages, tools);
  let lastError;
  for (let attempt = 0; attempt <= (options.retries ?? 2); attempt += 1) {
    try {
      const raw = await transport(settings.apiUrl, { method: 'POST', headers: authHeaders(settings), body: JSON.stringify(payload) }, settings.apiTimeoutMs);
      return backend.normalizeResponse(raw);
    } catch (error) {
      lastError = error;
      if (!(error instanceof RetryableProviderError) || attempt >= (options.retries ?? 2)) break;
      await sleep(2 ** attempt * 100);
    }
  }
  throw new ProviderError(`Provider request failed after ${(options.retries ?? 2) + 1} attempts: ${lastError?.message ?? lastError}`);
}

module.exports = {
  BUILTIN_PROTOCOLS,
  ProviderError,
  RetryableProviderError,
  availableBackends,
  backendFor,
  complete,
  registerBackend,
  valueAtPath,
};
